pkg/controllers/nodeclass/status/images.go

```diff
@@ -147,9 +173,50 @@ func imageVersionsUnready(nodeClass *v1alpha2.AKSNodeClass) bool {
 }
 
 // Handles case 4: check if the maintenance window is open
-func isMaintenanceWindowOpen() bool {
-	// TODO: need to add the actual logic for handling maintenance windows once it is in the ConfigMap.
-	return true
+// TODO (charliedmcb): remove nolint on gocyclo. Added for now in order to pass "make verify"
+// I think the best way to get rid of gocyclo is to break the section retrieving the maintenance window
+// range from the ConfigMap into its own helper function using channel as a parameter.
+// nolint: gocyclo
+func (r *NodeImageReconciler) isMaintenanceWindowOpen(ctx context.Context) (bool, error) {
+	if r.systemNamespace == "" {
+		// We fail open here, since the default case should be to upgrade
+		return true, nil
+	}
+
+	mwConfigMap, err := r.inClusterKubernetesInterface.CoreV1().ConfigMaps(r.systemNamespace).Get(ctx, maintenanceWindowConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// We fail open here, since the default case should be to upgrade
+			return true, nil
+		}
+		return false, fmt.Errorf("error getting maintenance window configmap, %w", err)
+	}
+	if len(mwConfigMap.Data) == 0 {
+		// An empty configmap means there's no maintenance windows defined, and its up to us when to preform maintenance
+		return true, nil
+	}
+
+	nextNodeOSMWStartStr, okStart := mwConfigMap.Data[fmt.Sprintf(configMapStartTimeFormat, nodeOSMaintenanceWindowChannel)]
+	nextNodeOSMWEndStr, okEnd := mwConfigMap.Data[fmt.Sprintf(configMapEndTimeFormat, nodeOSMaintenanceWindowChannel)]
+	if !okStart && !okEnd {
+		// No maintenance window defined for aksManagedNodeOSUpgradeSchedule, so its up to us when to preform maintenance
+		return true, nil
+	} else if (okStart && !okEnd) || (!okStart && okEnd) {
+		return false, fmt.Errorf("unexpected state, with incomplete maintenance window data for channel %s", nodeOSMaintenanceWindowChannel)
+	}
+
+	nextNodeOSMWStart, err := time.Parse(time.RFC3339, nextNodeOSMWStartStr)
+	if err != nil {
+		return false, fmt.Errorf("error parsing maintenance window start time for channel %s, %w", nodeOSMaintenanceWindowChannel, err)
+	}
+	nextNodeOSMWEnd, err := time.Parse(time.RFC3339, nextNodeOSMWEndStr)
+	if err != nil {
+		return false, fmt.Errorf("error parsing maintenance window end time for channel %s, %w", nodeOSMaintenanceWindowChannel, err)
+	}
+
+	now := time.Now().UTC()
```

Comment:
minor: A trick I've often seen used, and definitely recommend, is having some sort of `TimeProvider` rather than just reading system time directly here.

Probably not worth fixing here because we call time.Now() in other places too, but generally being able to swap your time provider out for unit testing can be very helpful. Curious what @tallaxes thinks about this... maybe a low pri future improvement...